runtime/src/accounts_db.rs

```diff
@@ -879,6 +881,144 @@ impl AccountsDB {
+            let no_ancestors = HashMap::new();
+            let accounts_index = self.accounts_index.read().unwrap();
+            stored_accounts
+                .iter()
+                .filter(
+                    |(
+                        pubkey,
+                        _account,
+                        _hash,
+                        _storage_size,
+                        (store_id, offset),
+                        _write_version,
+                    )| {
+                        if let Some((list, _)) = accounts_index.get(pubkey, &no_ancestors) {
+                            list.iter()
+                                .any(|(_slot, i)| i.store_id == *store_id && i.offset == *offset)
+                        } else {
+                            false
+                        }
+                    },
+                )
+                .collect()
+        };
+
+        let alive_total: u64 = alive_accounts
+            .iter()
+            .fold(0, |t, (_, _, _, a, _, _)| t + *a as u64);
+        let aligned_total: u64 = (alive_total + (PAGE_SIZE - 1)) & !(PAGE_SIZE - 1);
+
+        debug!(
+            "shrinking: slot: {}, stored_accounts: {} => alive_accounts: {} ({} bytes; aligned to: {})",
+            slot,
+            stored_accounts.len(),
+            alive_accounts.len(),
+            alive_total,
+            aligned_total
+        );
+
+        if aligned_total > 0 {
+            let mut accounts = Vec::with_capacity(alive_accounts.len());
+            let mut hashes = Vec::with_capacity(alive_accounts.len());
+            let mut write_versions = Vec::with_capacity(alive_accounts.len());
+
+            for (pubkey, account, hash, _size, _location, write_version) in alive_accounts {
+                accounts.push((pubkey, account));
+                hashes.push(*hash);
+                write_versions.push(*write_version);
+            }
+
+            let shrank_store = self.create_and_insert_store(slot, aligned_total);
+            let infos = self.store_accounts_to(
+                slot,
+                &accounts,
+                &hashes,
+                |_| shrank_store.clone(),
+                write_versions.into_iter(),
+            );
+            let reclaims = self.update_index(slot, infos, &accounts);
+
+            self.handle_reclaims(&reclaims);
+
+            let mut storage = self.storage.write().unwrap();
+            if let Some(slot_storage) = storage.0.get_mut(&slot) {
+                slot_storage.retain(|_key, store| store.count() > 0);
+            }
+        }
+    }
+
+    fn next_shrink_slot(&self) -> Option<Slot> {
+        let mut candidates = self.shrink_candidate_slots.lock().unwrap();
+        let next = candidates.pop();
+        if next.is_some() {
+            next
+        } else {
+            let mut new_all_slots = self.all_root_slots_in_index();
+            let next = new_all_slots.pop();
+            std::mem::replace(&mut *candidates, new_all_slots);
```

Comment:
would prefer to take the shink_candiate_slot lock twice here to avoid taking two locks at once

Reply:
yeah, nice catch!